cmd/consumer/main.go

```diff
@@ -0,0 +1,230 @@
+package main
+
+import (
+	"context"
+	"fmt"
+	"log"
+	"os"
+	"os/signal"
+	"runtime"
+	"sync"
+	"time"
+
+	"github.com/alexdunne/gs-onboarding/hn"
+	"github.com/alexdunne/gs-onboarding/internal/postgres"
+	"github.com/pkg/errors"
+	"github.com/spf13/viper"
+	"go.uber.org/zap"
+)
+
+type Config struct {
+	WorkerCount            int
+	WorkerIntervalDuration time.Duration
+	DatabaseDSN            string
+}
+
+func loadConfig() (*Config, error) {
+	viper.SetConfigFile(".env")
+	if err := viper.ReadInConfig(); err != nil {
+		return nil, errors.Wrap(err, "failed to read env file")
+	}
+
+	c := &Config{
+		WorkerCount:            runtime.NumCPU(),
+		WorkerIntervalDuration: 300 * time.Second,
+		DatabaseDSN: fmt.Sprintf(
+			"postgres://%s:%s@%s:%s/%s",
+			viper.GetString("DATABASE_USER"),
+			viper.GetString("DATABASE_PASSWORD"),
+			viper.GetString("DATABASE_HOST"),
+			viper.GetString("DATABASE_PORT"),
+			viper.GetString("DATABASE_DB"),
+		),
+	}
+
+	intervalSeconds := viper.GetInt("WORKER_INTERVAL_SECONDS")
+	if intervalSeconds != 0 {
+		c.WorkerIntervalDuration = time.Duration(intervalSeconds) * time.Second
+	}
+
+	return c, nil
+}
+
+func main() {
+	if err := run(); err != nil {
+		log.Fatal(err)
+		os.Exit(1)
+	}
+}
+
+func run() error {
+	ctx, cancel := context.WithCancel(context.Background())
+	defer cancel()
+
+	cfg, err := loadConfig()
+	if err != nil {
+		return errors.Wrap(err, "loading config")
+	}
+
+	logger, err := zap.NewProduction()
+	if err != nil {
+		return errors.Wrap(err, "creating logger")
+	}
+	defer logger.Sync()
+
+	go func() {
+		// run the consumer the first time when the goroutine starts
+		consumer(ctx, cfg, logger)
```

Comment:
Realistically you probably wouldn't want to call this here.

You wont get a tick on 0 and a slightly longer boot time but you run the risk of doubling your consumers up.

Reply:
Makes sense. After staring at it for a bit I realised I could wrap it in a `for` so `consumer` is only called once. e.g.:

```
go func() {
		for {
			logger.Info("running consumer")
			err := consumer.run(ctx)
			if err != nil {
				logger.Error("consumer failed", zap.Error(err))
			}

			select {
			case <-ctx.Done():
				return
			case <-time.Tick(cfg.WorkerIntervalDuration):
			}
		}
	}()
```